compiler/optimizer/src/main/java/org/apache/nemo/compiler/optimizer/pass/compiletime/reshaping/SamplingTaskSizingPass.java

```diff
@@ -0,0 +1,468 @@
+      stageStartingVertices, verticesWithStageOutgoingEdges, stageEndingVertices, partitionerProperty);
+    // By default, set the number of iterations as 2
+    toInsert.setMaxNumberOfIterations(2);
+    // make edges connected to splitter vertex
+    final Set<IREdge> fromOutsideToSplitter = setEdgesFromOutsideToSplitter(dag, toInsert, stageStartingVertices);
+    final Set<IREdge> fromSplitterToOutside = setEdgesFromSplitterToOutside(dag, toInsert,
+      verticesWithStageOutgoingEdges);
+    final Set<IREdge> edgesWithSplitterVertex = new HashSet<>();
+    edgesWithSplitterVertex.addAll(fromOutsideToSplitter);
+    edgesWithSplitterVertex.addAll(fromSplitterToOutside);
+    //fill in splitter vertex information
+    toInsert.insertWorkingVertices(stageVertices, edgesBetweenOriginalVertices);
+
+    //map splitter vertex connection to corresponding internal vertex connection
+    for (IREdge splitterEdge : fromSplitterToOutside) {
+      for (IREdge internalEdge : fromOriginalToOutside) {
+        if (splitterEdge.getDst() instanceof TaskSizeSplitterVertex) {
+          TaskSizeSplitterVertex nextSplitter = (TaskSizeSplitterVertex) splitterEdge.getDst();
+          if (nextSplitter.getOriginalVertices().contains(internalEdge.getDst())) {
+            toInsert.mapEdgeWithLoop(splitterEdge, internalEdge);
+          }
+        } else {
+          if (splitterEdge.getDst().equals(internalEdge.getDst())) {
+            toInsert.mapEdgeWithLoop(splitterEdge, internalEdge);
+          }
+        }
+      }
+    }
+    for (IREdge splitterEdge : fromOutsideToSplitter) {
+      for (IREdge internalEdge : fromOutsideToOriginal) {
+        if (splitterEdge.getSrc().equals(internalEdge.getSrc())) {
+          toInsert.mapEdgeWithLoop(splitterEdge, internalEdge);
+        }
+      }
+    }
+
+    final SignalVertex signalVertex = new SignalVertex();
+    fromOutsideToOriginal.forEach(toInsert::addDagIncomingEdge);
+    fromOutsideToOriginal.forEach(toInsert::addNonIterativeIncomingEdge); //
+    fromOriginalToOutside.forEach(toInsert::addDagOutgoingEdge);
+    toInsert.insertSignalVertex(signalVertex);
+
+    // insert splitter vertex
+    dag.insert(toInsert, incomingEdgesOfOriginalVertices, outgoingEdgesOfOriginalVertices,
+      edgesWithSplitterVertex);
+
+    toInsert.printLogs();
+  }
+
+  /**
+   * Changes stage outgoing edges' execution property from one-to-one to shuffle when stage incoming edge became the
+   * target of DTS.
+   * Need to be careful about referenceShuffleEdge because this code does not check whether it is a valid shuffle edge
+   * or not.
+   * @param edge                   edge to change execution property.
+   * @param referenceShuffleEdge  reference shuffle edge to copy key related execution properties
+   * @param partitionerProperty   partitioner property of shuffle
+   */
+  private IREdge changeOneToOneEdgeToShuffleEdge(final IREdge edge,
```

Comment:
This method has no usages

Reply:
This method was created to enable inserting splitter vertex to stages which has non-shuffle stage edges.

For some reason, changing communication property manually in Nemo encounters runtime errors. I will leave this method as TODO and add some explanations.

Reply:
Sounds good. Thanks!

Reply:
I've added a TODO related to this problem.

Please check:
https://issues.apache.org/jira/browse/NEMO-452